consensus/scheduled_task.go

```diff
@@ -0,0 +1,61 @@
+package consensus
+
+import (
+	"sync"
+	"time"
+)
+
+// ScheduledTask is a thread-safe wrapper around time.Timer that allows for creating tasks that
+// can be scheduled to execute at a later time with pre-specified params. Both the params the
+// task are fully defined at the time of scheduling. Once a task has been scheduled, it cannot
+// be modified. However, tasks can be cancelled and rescheduled.
+//
+// This pattern is useful for spawning off tasks that we want to run after some specified amount
+// of time, but still want to have the ability to cancel.
+type ScheduledTask[TaskParam any] struct {
+	lock     sync.RWMutex
+	timer    *time.Timer
+	duration time.Duration
+}
+
+func NewScheduledTask[TaskParam any]() *ScheduledTask[TaskParam] {
+	return &ScheduledTask[TaskParam]{
+		lock:  sync.RWMutex{},
+		timer: nil,
+	}
+}
+
+// Cancel the currently scheduled task, if any, and schedule a new task to be executed after the countdown duration.
```

Comment:
should this comment be next to the cancel function?

Reply:
Nah, this is the correct place. I've reworded the comment so it's a bit cleraer